File: day7.pl
#!/usr/bin/env perl

# perl solution for day 7 part 1
use strict;
use warnings;

use File::Slurp;

sub found_abba {
  return 1 if $_[0] =~ /(.)(?!\1)(.)\2\1/;
  return 0;
}

sub supports_tls {
  my $input = $_[0];
  my $tls = found_abba($input);

  while ($input =~ /(\[.*?\])/g) {
    if (found_abba($1)) {
      $tls = 0;
    }
  }
  return $tls;
}

my @ips = read_file('input.txt');
my $total = 0;

foreach my $ip (@ips) {
  if (supports_tls($ip)) {
    $total += 1;
  }
}

print "$total\n"

File: day7.py
#!/usr/bin/env python3

myfile = open('input.txt', 'r')
contents = myfile.read()
myfile.close()

contents = contents.strip()
contents = contents.split()

total = 0

for i in contents:
    paran = False
    supports = False
    supports2 = False
    inside = set()
    outside = set()

    for j in range(len(i) - 2):
        if i[j] == '[':
            paran = True
        if i[j] == ']':
            paran = False
        if i[j] == i[j+2] and i[j] != i[j+1]:
            if not paran:
                if i[j] + i[j+1] in inside:
                    supports = True
                    break
                outside.add(i[j] + i[j +1])
            else:
                if i[j + 1] + i[j] in outside:
                    supports = True
                    break
                inside.add(i[j+1] + i[j])

    if supports:
        total += 1

print(total)
